### PR TITLE
cache mutation detector: improve diff readability

### PR DESCRIPTION
**Pull request by enj**
This change updates the struct diff logic used by the cache mutation
detector to make it easier to see what has changed between the two
structs.  It also makes the mutation detector resilient against code
that modifies objects temporarily before resetting them back into
their original state (previously this could lead to diff output that
stated that the two objects were identical).

Signed-off-by: Monis Khan <mkhan@redhat.com>

/kind bug
/priority important-soon

```release-note
NONE
```

this PR:

```
$ go test -v k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache -run ^TestMutationDetector$
=== RUN   TestMutationDetector
CACHE name[0] ALTERED!
  &v1.Pod{
  	TypeMeta: v1.TypeMeta{},
  	ObjectMeta: v1.ObjectMeta{
  		... // 8 identical fields
  		DeletionTimestamp:          nil,
  		DeletionGracePeriodSeconds: nil,
  		Labels: map[string]string{
+ 			"change": "true",
  			"check":  "foo",
  		},
  		Annotations:     nil,
  		OwnerReferences: nil,
  		... // 4 identical fields
  	},
  	Spec:   v1.PodSpec{},
  	Status: v1.PodStatus{},
  }

--- PASS: TestMutationDetector (1.00s)
PASS
ok  	k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache
```

master:

```
$ go test -v k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache -run ^TestMutationDetector$
=== RUN   TestMutationDetector
CACHE name[0] ALTERED!
(*v1.Pod)(0xc000198800)({                              (*v1.Pod)(0xc000199000)({
 TypeMeta: (v1.TypeMeta) {                              TypeMeta: (v1.TypeMeta) {
  Kind: (string) "",                                     Kind: (string) "",
  APIVersion: (string) ""                                APIVersion: (string) ""
 },                                                     },
 ObjectMeta: (v1.ObjectMeta) {                          ObjectMeta: (v1.ObjectMeta) {
  Name: (string) (len=8) "anything",                     Name: (string) (len=8) "anything",
  GenerateName: (string) "",                             GenerateName: (string) "",
  Namespace: (string) "",                                Namespace: (string) "",
  SelfLink: (string) "",                                 SelfLink: (string) "",
  UID: (types.UID) "",                                   UID: (types.UID) "",
  ResourceVersion: (string) "",                          ResourceVersion: (string) "",
  Generation: (int64) 0,                                 Generation: (int64) 0,
  CreationTimestamp: (v1.Time) {                         CreationTimestamp: (v1.Time) {
   Time: (time.Time) {                                    Time: (time.Time) {
    wall: (uint64) 0,                                      wall: (uint64) 0,
    ext: (int64) 0,                                        ext: (int64) 0,
    loc: (*time.Location)(<nil>)                           loc: (*time.Location)(<nil>)
   }                                                      }
  },                                                     },
  DeletionTimestamp: (*v1.Time)(<nil>),                  DeletionTimestamp: (*v1.Time)(<nil>),
  DeletionGracePeriodSeconds: (*int64)(<nil>),           DeletionGracePeriodSeconds: (*int64)(<nil>),
  Labels: (map[string]string) (len=2) {                  Labels: (map[string]string) (len=1) {
   (string) (len=5) "check": (string) (len=3) "foo",      (string) (len=5) "check": (string) (len=3) "foo"
   (string) (len=6) "change": (string) (len=4) "true"    },
  },                                                     Annotations: (map[string]string) <nil>,
  Annotations: (map[string]string) <nil>,                OwnerReferences: ([]v1.OwnerReference) <nil>,
  OwnerReferences: ([]v1.OwnerReference) <nil>,          Initializers: (*v1.Initializers)(<nil>),
  Initializers: (*v1.Initializers)(<nil>),               Finalizers: ([]string) <nil>,
  Finalizers: ([]string) <nil>,                          ClusterName: (string) "",
  ClusterName: (string) "",                              ManagedFields: ([]v1.ManagedFieldsEntry) <nil>
  ManagedFields: ([]v1.ManagedFieldsEntry) <nil>        },
 },                                                     Spec: (v1.PodSpec) {
 Spec: (v1.PodSpec) {                                    Volumes: ([]v1.Volume) <nil>,
  Volumes: ([]v1.Volume) <nil>,                          InitContainers: ([]v1.Container) <nil>,
  InitContainers: ([]v1.Container) <nil>,                Containers: ([]v1.Container) <nil>,
  Containers: ([]v1.Container) <nil>,                    RestartPolicy: (v1.RestartPolicy) "",
  RestartPolicy: (v1.RestartPolicy) "",                  TerminationGracePeriodSeconds: (*int64)(<nil>),
  TerminationGracePeriodSeconds: (*int64)(<nil>),        ActiveDeadlineSeconds: (*int64)(<nil>),
  ActiveDeadlineSeconds: (*int64)(<nil>),                DNSPolicy: (v1.DNSPolicy) "",
  DNSPolicy: (v1.DNSPolicy) "",                          NodeSelector: (map[string]string) <nil>,
  NodeSelector: (map[string]string) <nil>,               ServiceAccountName: (string) "",
  ServiceAccountName: (string) "",                       DeprecatedServiceAccount: (string) "",
  DeprecatedServiceAccount: (string) "",                 AutomountServiceAccountToken: (*bool)(<nil>),
  AutomountServiceAccountToken: (*bool)(<nil>),          NodeName: (string) "",
  NodeName: (string) "",                                 HostNetwork: (bool) false,
  HostNetwork: (bool) false,                             HostPID: (bool) false,
  HostPID: (bool) false,                                 HostIPC: (bool) false,
  HostIPC: (bool) false,                                 ShareProcessNamespace: (*bool)(<nil>),
  ShareProcessNamespace: (*bool)(<nil>),                 SecurityContext: (*v1.PodSecurityContext)(<nil>),
  SecurityContext: (*v1.PodSecurityContext)(<nil>),      ImagePullSecrets: ([]v1.LocalObjectReference) <nil>,
  ImagePullSecrets: ([]v1.LocalObjectReference) <nil>,   Hostname: (string) "",
  Hostname: (string) "",                                 Subdomain: (string) "",
  Subdomain: (string) "",                                Affinity: (*v1.Affinity)(<nil>),
  Affinity: (*v1.Affinity)(<nil>),                       SchedulerName: (string) "",
  SchedulerName: (string) "",                            Tolerations: ([]v1.Toleration) <nil>,
  Tolerations: ([]v1.Toleration) <nil>,                  HostAliases: ([]v1.HostAlias) <nil>,
  HostAliases: ([]v1.HostAlias) <nil>,                   PriorityClassName: (string) "",
  PriorityClassName: (string) "",                        Priority: (*int32)(<nil>),
  Priority: (*int32)(<nil>),                             DNSConfig: (*v1.PodDNSConfig)(<nil>),
  DNSConfig: (*v1.PodDNSConfig)(<nil>),                  ReadinessGates: ([]v1.PodReadinessGate) <nil>,
  ReadinessGates: ([]v1.PodReadinessGate) <nil>,         RuntimeClassName: (*string)(<nil>),
  RuntimeClassName: (*string)(<nil>),                    EnableServiceLinks: (*bool)(<nil>),
  EnableServiceLinks: (*bool)(<nil>),                    PreemptionPolicy: (*v1.PreemptionPolicy)(<nil>),
  PreemptionPolicy: (*v1.PreemptionPolicy)(<nil>),       Overhead: (v1.ResourceList) <nil>
  Overhead: (v1.ResourceList) <nil>                     },
 },                                                     Status: (v1.PodStatus) {
 Status: (v1.PodStatus) {                                Phase: (v1.PodPhase) "",
  Phase: (v1.PodPhase) "",                               Conditions: ([]v1.PodCondition) <nil>,
  Conditions: ([]v1.PodCondition) <nil>,                 Message: (string) "",
  Message: (string) "",                                  Reason: (string) "",
  Reason: (string) "",                                   NominatedNodeName: (string) "",
  NominatedNodeName: (string) "",                        HostIP: (string) "",
  HostIP: (string) "",                                   PodIP: (string) "",
  PodIP: (string) "",                                    StartTime: (*v1.Time)(<nil>),
  StartTime: (*v1.Time)(<nil>),                          InitContainerStatuses: ([]v1.ContainerStatus) <nil>,
  InitContainerStatuses: ([]v1.ContainerStatus) <nil>,   ContainerStatuses: ([]v1.ContainerStatus) <nil>,
  ContainerStatuses: ([]v1.ContainerStatus) <nil>,       QOSClass: (v1.PodQOSClass) ""
  QOSClass: (v1.PodQOSClass) ""                         }
 }                                                     })
})                                                     
                                                       

--- PASS: TestMutationDetector (1.00s)
PASS
ok  	k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache	1.010s
```

/assign @caesarxuchao @sttts @deads2k 
xref: #78869

### CODE REVIEW

File: staging/src/k8s.io/client-go/go.mod [hunk 1]
```diff
@@ -13,6 +13,7 @@ require (
 	github.com/golang/groupcache v0.0.0-20160516000752-02826c3e7903
 	github.com/golang/protobuf v1.2.0
 	github.com/google/btree v0.0.0-20160524151835-7d79101e329e // indirect
+	github.com/google/go-cmp v0.3.0
```
Comment (enj) [inline staging/src/k8s.io/client-go/go.mod]: Note that I could "avoid" this direct dependency by continuing to use the deprecated functions from the `k8s.io/apimachinery/pkg/util/diff` package (but that looks like the wrong approach):

```go
func legacyDiff(a, b interface{}) string {
	return cmp.Diff(a, b)
}

// ObjectDiff prints the diff of two go objects and fails if the objects
// contain unhandled unexported fields.
// DEPRECATED: use github.com/google/go-cmp/cmp.Diff
func ObjectDiff(a, b interface{}) string {
	return legacyDiff(a, b)
}
```

File: staging/src/k8s.io/client-go/tools/cache/mutation_detector.go [hunk 2]
```diff
@@ -113,8 +113,12 @@ func (d *defaultCacheMutationDetector) CompareObjects() {
 
 	altered := false
 	for i, obj := range d.cachedObjs {
-		if !reflect.DeepEqual(obj.cached, obj.copied) {
-			fmt.Printf("CACHE %s[%d] ALTERED!\n%v\n", d.name, i, diff.ObjectGoPrintSideBySide(obj.cached, obj.copied))
+		// make sure that we have a stable view of the cached object
+		// sometimes code will modify an object and then reset it back to the original state
+		// this can result in reflect.DeepEqual returning false while cmp.Diff prints an empty diff
+		cached := obj.cached.DeepCopyObject()
+		if !reflect.DeepEqual(cached, obj.copied) {
```
Comment (sttts) [inline staging/src/k8s.io/client-go/tools/cache/mutation_detector.go]: this is even too weak as an equality. We want pointer-equality, not just value equality.

Comment (sttts) [reply staging/src/k8s.io/client-go/tools/cache/mutation_detector.go]: We compare against a deep copy. Pointers won't be equal. But there is a chance that we miss mutations if they change pointers to extensionally equal data. But not much we can do, and that restriction was here already before the PR.

File: staging/src/k8s.io/client-go/tools/cache/mutation_detector.go [hunk 3]
```diff
@@ -113,8 +113,12 @@ func (d *defaultCacheMutationDetector) CompareObjects() {
...
+			fmt.Printf("CACHE %s[%d] ALTERED!\n%v\n", d.name, i, cmp.Diff(obj.copied, cached))
```
Comment (sttts) [inline staging/src/k8s.io/client-go/tools/cache/mutation_detector.go]: `cmp.Diff` won't work against objects with private fields. I know we have them in URLs in JSONSchemaProps for CRDs, not sure anywhere else. `cmp.Diff` will panic.

Comment (sttts) [reply staging/src/k8s.io/client-go/tools/cache/mutation_detector.go]: we might want to catch panics here and do JSON string diff instead in those cases.

Comment (enj) [reply staging/src/k8s.io/client-go/tools/cache/mutation_detector.go]: We have API objects (runtime.Object) that are serialized on the wire that have private fields?

**Comment by yliaog** [comment]
please add a test

**Comment by enj** [comment]
> please add a test

A test that tests what functionality?

**Comment by yliaog** [comment]
i'm thinking a test to assert that the code can achieve "It also makes the mutation detector resilient against code that modifies objects temporarily before resetting them back into their original state"

**Pull request closed without merging**